src/main/java/org/jenkinsci/plugins/workflow/support/DefaultStepContext.java

```diff
@@ -97,6 +99,33 @@
         }
     }
 
+    private synchronized TaskListener getListener() throws IOException, InterruptedException {
+        if (listener == null) {
+            FlowNode node = getNode();
+            if (!node.isActive()) {
+                throw new IOException("cannot start writing logs to a finished node " + node + " " + node.getDisplayFunctionName() + " in " + node.getExecution());
+            }
+            listener = LogStorageAction.listenerFor(node, get(ConsoleLogFilter.class));
+            LOGGER.log(Level.FINE, "opened log for {0}", node.getDisplayFunctionName());
+            if (listener instanceof AutoCloseable) {
+                node.getExecution().addListener(new GraphListener.Synchronous() {
+                    @Override public void onNewHead(FlowNode newNode) {
+                        if (!node.isActive()) {
+                            node.getExecution().removeListener(this);
+                            LOGGER.log(Level.FINE, "closing log for {0}", node.getDisplayFunctionName());
```

Comment:
For a good time, try to disentangle what all the xxFunctionName/xxDisplayName function for a FlowNode do and which to use when. 

And by a "good time" I mean "if you're a masochist."